tools/federation/src/main/java/org/eclipse/rdf4j/federated/repository/FedXRepositoryConfig.java

```diff
@@ -205,4 +292,120 @@ public void parse(Model m, Resource implNode) throws RepositoryConfigException {
+		if (getConfig() == null) {
+			setConfig(new FedXConfig());
+		}
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_JOIN_WORKER_THREADS, null))
+				.ifPresent(value -> config.withJoinWorkerThreads(value.intValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_UNION_WORKER_THREADS, null))
+				.ifPresent(value -> config.withUnionWorkerThreads(value.intValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_LEFT_JOIN_WORKER_THREADS, null))
+				.ifPresent(value -> config.withLeftJoinWorkerThreads(value.intValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_BOUND_JOIN_BLOCK_SIZE, null))
+				.ifPresent(value -> config.withBoundJoinBlockSize(value.intValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_ENFORCE_MAX_QUERY_TIME, null))
+				.ifPresent(value -> config.withEnforceMaxQueryTime(value.intValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_ENABLE_SERVICE_AS_BOUND_JOIN, null))
+				.ifPresent(value -> config.withEnableServiceAsBoundJoin(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_ENABLE_MONITORING, null))
+				.ifPresent(value -> config.withEnableMonitoring(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_LOG_QUERY_PLAN, null))
+				.ifPresent(value -> config.withLogQueryPlan(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_LOG_QUERIES, null))
+				.ifPresent(value -> config.withLogQueries(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_DEBUG_QUERY_PLAN, null))
+				.ifPresent(value -> config.withDebugQueryPlan(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_INCLUDE_INFERRED_DEFAULT, null))
+				.ifPresent(value -> config.withIncludeInferredDefault(value.booleanValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_SOURCE_SELECTION_CACHE_SPEC, null))
+				.ifPresent(value -> config.withSourceSelectionCacheSpec(value.stringValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_PREFIX_DECLARATIONS, null))
+				.ifPresent(value -> config.withPrefixDeclarations(value.stringValue()));
+
+		Models.objectLiteral(m.getStatements(confNode, CONFIG_CONSUMING_ITERATION_MAX, null))
+				.ifPresent(value -> config.withConsumingIterationMax(value.intValue()));
+
+	}
+
+	/**
+	 * Export the provided {@link FedXConfig} to its RDF representation. Note that {@link #getConfig()} could be null if
+	 * configuration has not been set yet.
+	 *
+	 * @param config   the configuration to export
+	 * @param implNode the node to which to write the config reference (i.e. {@link #FEDX_CONFIG}) to
+	 */
+	protected void exportFedXConfig(Model model, Resource implNode) {
+		if (getConfig() == null) {
+			return;
+		}
+
+		BNode confNode = Values.bnode();
+
+		model.add(confNode, CONFIG_JOIN_WORKER_THREADS, vf.createLiteral(config.getJoinWorkerThreads()));
```

Comment:
the downside of the current implementation is that exporting will also write the FedXConfig default values, i.e. values that may not have been present in the repository configuration ttl before.

Ideally the parse / export logic would be consistent, i.e. running parse and export produces the same result.

I think I could live with this, but I would like hear @hmottestad's opinion whether this is in line with the RDF4J design and other repo configurations.

Reply:
> Ideally the parse / export logic would be consistent, i.e. running parse and export produces the same result.

But isn't the issue here that we don't know for sure whether the user explicitly set values. Also, isn't being explicit better than implicit?
(Otherwise I suppose one could, for each individual value, check the default first)

Reply:
I would leave it up to @hmottestad as he has likely the broadest overview at this point.

As a side note: we are usually solving the parsing / export consistency by having "null" state for the field (i.e. Boolean, Integer, ...). In the current design this does not really fit as it uses the API level FedXConfig. To achieve this properly something similar one would need a repository level FedXConfig and some translation to the api level FedXConfig.